rfpremier/ah-django
# #164857224 user notifications

### What does this PR do?
This PR adds the send notifications to a user functionality via
1. In-app notifications
2. Email notifications

Notifications are sent when
1. An author creates a new article then the notifications of the article are sent to the author's followers
2. When a new comment is created on an article then the authors that favourited the article get a new notification on the comment created.
3. When a user follows another user.

### Tasks to be completed?

  - [x]  - Initialized notifications app
  
  - [x]  - Initialize test dir & get rid of unnecessary files
    
   - [x] - Initialize the signal to capture event once an article is saved
   - [x] - Add email notification & in app notification fields to user model
    
  - [x]  - Get all user followers
    
   - [x] - create send mail functionality
    
   - [x] - create send notification functionality
   
 - [x] - create custom middleware to send the request to models & signals

 - [x] - create mail template
           
 - [x] - make small refactors
 
 - [x] - rebased from highlited comments PR

 - [x] - create send email and in app notifications functionality & hook them
           up with Article and Comment signals

 - [x] - initialized view file

- [x] - add tests for opt in and out views
    
- [x] - add opt in and out functionality

- [x] -  receive new notification when a user gets a new follower.
    
    
### How can this be manually tested?
1. Clone this repo.
2. Checkout to the `ft/164857224-user-notifications`
3. Run `python manage.py makemigrations` , `python manage.py migrate` & `python manage.py runserver`
4. Create 2 new user via the `POST http://127.0.0.1:8000/api/users/`
You can decide to replace the second user email with your own just to see the mails hitting your mail box.

```js
{
   "user": {
  	 "email": "first@gmail.com",
     "password": "first2019",
     "username": "first"
    }
 }
```

```js
{
    "user": {
        "email": "profilemania@gmail.com",
        "password": "profiles2019",
        "username": "profmania"
    }
}
```
5. Open up a websocket client 
[Install this one if you hadn't already](https://chrome.google.com/webstore/detail/browser-websocket-client/mdmlhchldhfnfnkfmljgeinlffmdgkjo?utm_source=chrome-ntp-icon)

6. Paste in this link in the websocket client and then connect.
```js
ws://127.0.0.1:8000/api/notifications/first/
``` 
7. Let the second user follow the first 1.
Place the second user's auth token in the Authorization headers

`POST http://127.0.0.1:8000/api/users/first/follow/`

8. Open the websocket client and you will see a new notification

```js
{
  "type": "notify",
  "message": "profmania has started following you"
}
``` 

![image](https://user-images.githubusercontent.com/12128153/56535543-94da9a00-6564-11e9-9236-7337b500238f.png)

this shows that you receive a notification once a user follows you.


9. Paste in this link in the websocket client and then connect.
```js
ws://127.0.0.1:8000/api/notifications/profmania/
``` 
We are pasting using the username of the second user.
N/B:
Details of the design decision are covered below in the context section.

10. Create a new article as the first user.
   `POST http://127.0.0.1:8000/api/articles`
N/B: Remember to have the first user's token in the Authorization field
```js
{
    "article": {
        "title": "This is the article with notifications",
        "description": "This is the article description",
        "body": "This is the article body",
        "image_url": "https://imageurl.com"
    }
}
```


11. Open up the client and see the new object returned in the received messages section.
```js
{
  "type": "notify",
  "message": "New article from first named This is the article with notifications"
}
```
If you had used your own credentials as the second user, check your email inbox. 
This shows that once a user follows another, they always receive notifications on the articles the users they follow have published by default.

12. Create a third user.
`POST http://127.0.0.1:8000/api/users/`

```js
{
    "user": {
        "email": "third@gmail.com",
        "password": "third2019",
        "username": "third"
    }
}
```
13. Let the third user favorite the article
```js
[POST] http://127.0.0.1:8000/api/articles/this-is-the-article-with-notifications/favorite
```
14. Open up the websocket client and connect to the third user's chanel
```js
ws://127.0.0.1:8000/api/notifications/third/
```
15. Let the second user create a comment in the article

```js
POST http://127.0.0.1:8000/api/articles/this-is-the-article-with-notifications/comments
```

```js
{
	"comment": {
		"body": "New comment"
	}
}
```

16. Open up the websocket client and you will see the following object in the received messages
```js
{
  "type": "notify",
  "message": "A New comment posted by profmania on article This is the article with notifications"
}
```

### Opting out of in app notifications.

17. Let the third user opt out of `in_app` notifications
N/B: Please use the third user's token in the auth headers.
```
[DELETE] http://127.0.0.1:8000/api/subscription/in_app/
```
You will get the following response

```js
{
    "data": "You have opted out of in_app notifications"
}
```
18. To see if this works, try creating a new comment as the second user again.

19. Check your websocket client  which will still be having this url connected
`ws://127.0.0.1:8000/api/notifications/third/`
and you will see that there is no new message received by the client.

### Opting out of email notifications
20.  Let the second user opt out of email notifications, hopefully you used your email credentials, if not kindly create a new user with your email credentials then follow the first user.
21. 
```
http://127.0.0.1:8000/api/subscription/email/
```
N/B: Use the second user's token
You will get the following response
```js
{
    "data": "You have opted out of email notifications"
}
```

22. Let the first user create another new article
```js
http://127.0.0.1:8000/api/articles/
```
N/B: Use the token of the first user
23. Note you will not receive any email concerning the new article.

### Any background context you want to add?

##### On sending notifications via websockets

Each user will be connected to their own websocket chanels via their username, meaning if your username was `johndoe` you will receive updates under the `/api/notifications/johndoe/` endpoint.
Updates include:
1. Receiving notifications on new articles from your followers.
2. Receiving notifications when articles you favorited get new comments.
3. Receiving notifications when you gain a new follower

Since the requests to test this functionality are many, [get the prepopulated endpoints here](https://www.getpostman.com/collections/5b7cdd4fc9c2cd85d943)

### Screenshots

1. On article creation
![image](https://user-images.githubusercontent.com/12128153/56502656-caa76080-651b-11e9-8dac-f680f888b8eb.png)

2. When an article you favourited gets a new comment.
![image](https://user-images.githubusercontent.com/12128153/56502690-ef033d00-651b-11e9-91de-3e7981b370d0.png)

3. When a user follows you

![image](https://user-images.githubusercontent.com/12128153/56533452-5347f000-6560-11e9-94b4-db11092f0f98.png)

![image](https://user-images.githubusercontent.com/12128153/56533664-ccdfde00-6560-11e9-8f86-27de847e0345.png)
